README.md
# pypowerup

[![Version](https://img.shields.io/pypi/v/pypowerup.svg)](https://pypi.python.org/pypi/pypowerup)
[![Build Status](https://img.shields.io/travis/sophiamyang/pypowerup.svg)](https://travis-ci.org/sophiamyang/pypowerup)
[![Docs](https://readthedocs.org/projects/pypowerup/badge/?version=latest)](https://pypowerup.readthedocs.io/en/latest/)

``pypowerup`` is the Python implementation for the research article "PowerUp!: A Tool for Calculating Minimum Detectable 
Effect Sizes and Minimum Required Sample Sizes for Experimental and Quasi-experimental Design Studies (Dong & Maynard, 2013)". It is a 
power analysis tool for experimental and quasi-experimental designs.



References
-------
Dong, N. & Maynard, R. A. (2013). PowerUp!: A tool for calculating minimum detectable
effect sizes and minimum required sample sizes for experimental and quasi- experimental design studies, Journal of Research on Educational Effectiveness, 6(1), 24-67. doi: 10.1080/19345747.2012.673143.
https://www.causalevaluation.org/uploads/7/3/3/6/73366257/powerup.xlsm

Bulus, M., Dong, N., Kelcey, B., & Spybrook, J. (2019). PowerUpR: Power Analysis Tools for Multilevel Randomized Experiments. R package version 1.0.4. https://CRAN.R-project.org/package=PowerUpR

HISTORY.md
# History

## 0.1.0 (2020-9-19)

* First release on PyPI.


pypowerup/tests/test_pypowerup.py
from pypowerup import effect_size, power, sample_size

def test_pypowerup():
    assert round(effect_size('ira', n=400), 2) == 0.28
    assert round(effect_size('ira', n=400, r21=0.5), 2) == 0.20
    assert sample_size('ira') == 787
    assert sample_size('ira', r21=0.8, g=1) == 159
    assert round(effect_size(design = "bira2_1c", n=80, J=14),2)== 0.17
    assert round(effect_size(design = "bira2_1c", n=80, J=14, r21=0.2, g=1),2)== 0.15
    assert sample_size(design = "bira2_1c", es=0.50, n=5, r21=0) == 26
    assert sample_size(design = "bira2_1c", es=0.40, n=5, r21=0.5) == 20
    assert round(effect_size(design = "bira2_1f", n=80, J=480),3) == 0.029
    assert round(effect_size(design = "bira2_1f", n=10, J=200, r21=0.05),3) == 0.122
    assert sample_size(design = "bira2_1f", es=0.50, n=10, r21=0, g=0)  == 13
    assert sample_size(design = "bira2_1f", es=0.50, n=6, r21=0.2, g=0)  == 17
    assert round(effect_size(design = "bira2_1r", n=80, J=480, rho2=0.35, omega2=0.1), 3) == 0.033
    assert round(effect_size(design = "bira2_1r", n=10, J=500, rho2=0.35, omega2=0.1), 3) == 0.068
    assert sample_size(design = "bira2_1r", n=10, g=1, r21=0.5, omega2=0.5, rho2=0.25, es=0.5) == 11
    assert sample_size(design = "bira2_1r", n=4, g=2, r21=0.5, omega2=0.5, rho2=0.25, es=0.5) == 18
    assert round(effect_size(design = "bira3_1r", n=80, J=10, K=100, rho3=0.2, rho2=0.15, omega3=0.1, omega2=0.1),3)==0.045
    assert round(effect_size(design = "bira3_1r", n=40, J=100, K=200, rho3=0.2, rho2=0.15, omega3=0.1, omega2=0.1),3)==0.029
    assert sample_size(design = "bira3_1r", es=0.15, n=20, J=30, rho3=0.1, rho2=0.15, omega3=0.1, omega2=0.1, r21=0.5) == 7
    assert sample_size(design = "bira3_1r", es=0.15, n=10, J=10, rho3=0.1, rho2=0.15, omega3=0.1, omega2=0.1, r21=0.5,r2t2=0,g=1) == 12
    assert round(effect_size(design = "bira4_1r", n=10, J=4, K=4, L=20, rho4=0.05, rho3=0.15, rho2=0.15, 
                omega4=0.5, omega3=0.5, omega2=0.5, r21=0.5, r2t2=0.5, r2t3=0.5, r2t4=0.5, g=1),3)==0.119
    assert round(effect_size(design = "bira4_1r", n=20, J=4, K=4, L=20, rho4=0.05, rho3=0.15, rho2=0.15, 
                omega4=0.5, omega3=0.5, omega2=0.5, r21=0.5, r2t2=0.5, r2t3=0.5, r2t4=0.5, g=1),3)==0.111
    assert sample_size(design = "bira4_1r", es=0.20, n=10, J=4, K=4, rho4=0.05, rho3=0.15, rho2=0.15, 
                omega4=0.5, omega3=0.5, omega2=0.5, r21=0.5, r2t2=0.5, r2t3=0.5, r2t4=0.5, g=1) == 9 
    assert sample_size(design = "bira4_1r", es=0.25, rho4=.05, rho3=.15, rho2=.15, omega4=.50, omega3=.50, omega2=.50, n=10, J=2, K=10) == 8
    assert round(power(design = "bira3_1r", rho3=.20, rho2=.15, omega3=.10, omega2=.10, n=69, J=10, K=100, es=.04,p=.50, r21=0, r2t2=0, r2t3=0, g=0),2)==0.70
    assert round(power(design = "bira4_1r", es=0.10, rho4=.05, rho3=.15, rho2=.15, omega4=.50, omega3=.50, omega2=.50, n=10, J=4, L=27, K=4),2)==0.50

    assert round(effect_size(design = "rd2_1f", n=55, J=20, r21=0.5, g=1, design_effect=2.75),3) == 0.198
    assert sample_size(design = "rd2_1f", n=20, es=0.20, r21=0.5, g=1, design_effect=2.75) == 54
    assert sample_size(design = "rd2_1f", n=20, es=0.10, r21=0.5, g=1, design_effect=2.75) == 216
    assert round(effect_size(design = "rd2_1r", n=50, J=40, r21=0.5, g=1, r2t2=0.1, omega2=0.2, rho2= 0.15, design_effect=2.75),3)==0.158
    assert sample_size(design = "rd2_1r", es=0.2, rho2=0.15, omega2=0.2, r21=0.5, r2t2=0.1, g=1, n=40, design_effect=2.75) == 30
    assert sample_size(design = "rd2_1r", es=0.1, rho2=0.15, omega2=0.2, r21=0.5, r2t2=0.1, g=1, n=60, design_effect=2.75) ==84
    assert round(effect_size(design = "rdc_2r", rho2=0.15, r21=0.5, r22=0.5, g=1, n=55, J=179, design_effect=2.75),3) ==0.201
    assert round(effect_size(design = "rdc_2r", rho2=0.15, r21=0.5, r22=0, g=1, n=55, J=200, design_effect=2.75),3) ==0.262
    assert sample_size(design = "rdc_2r", rho2=0.15, r21=0.5, r22=0.5, g=1, n=20, design_effect=2.75) == 210
    assert sample_size(design = "rdc_2r", rho2=0.15, r21=0.5, r22=0, g=1, n=200, design_effect=2.75) == 330
    assert round(effect_size(design = "rdc_3r", rho3=0.15, rho2=0.15, r21=0.5, r22=0.5, r23=0.5,  g=1, n=18, J=3, K=230, design_effect=2.75), 3) == 0.201
    assert sample_size(design = "rdc_3r", es=0.25, rho3=0.15, rho2=0.10, r21=0.5, r22=0.5, r23=0.5,  g=1, n=20, J=4, K=230, design_effect=2.75) == 129
    assert round(effect_size(design = "rd3_2f", rho2=0.15, r21=0.5, r22=0.5, g=0, n=18, J=3, K=71, design_effect=2.75),3) == 0.201
    assert round(effect_size(design = "rd3_2f", rho2=0.55, r21=0.3, r22=0.2, g=0, n=20, J=5, K=30, design_effect=2.75),3) == 0.516
    assert round(effect_size(design = "rd3_2r", rho3=0.15, rho2=0.15, omega3=1, r21=0.5, r22=0.5,r2t3=0.5, r23=0.5, g=0, K=80, n=18, J=3, design_effect=2.75),3) == 0.206
    assert sample_size(design = "rd3_2r", es=0.2, rho3=0.15, rho2=0.15, omega3=1, r21=0.5, r22=0.5,r2t3=0.5, r23=0.5, g=0, n=10, J=4, design_effect=2.75) == 76
    assert round(effect_size(design='cra2_2r', rho2=0.15, r21=0.40, r22=0.53, g=1, n=100, J=40), 3) == 0.250
    assert sample_size(design='cra2_2r', es=0.45, rho2=0.02, r21=0.01, r22=0.13, g=4, n=60, J=10) == 10
    assert round(effect_size(design='cra3_3r', rho3=0.38, rho2=0.10, r21=0.37, r22=0.53, r23=0.87, g=1, n=20, J=2, K=66), 3) ==0.199
    assert sample_size(design='cra3_3r', es=0.20, rho3=0.38, rho2=0.10, r21=0.37, r22=0.53, r23=0.87, g=1, n=20, J=2) == 66
    assert round(effect_size(design='cra4_4r', rho4=0.05, rho3=0.05, rho2=0.10, r21=0.50, r22=0.50, r23=0.50, r24=0.50, g=1, n=10, J=2, K=3, L=20),3) == 0.292
    assert sample_size(design='cra4_4r', es=0.20, rho4=0.05, rho3=0.05, rho2=0.10, r21=0.50, r22=0.50, r23=0.50, r24=0.50, g=1, n=5, J=2, K=3) == 45
    assert round(effect_size(design='bcra3_2f', rho2=0.10, r21=0.50, r22=0.50, g=1, n=20, J=44, K=5),3) == 0.102
    assert sample_size(design='bcra3_2f', es=0.15, rho2=0.30, r21=0.50, r22=0.50, g=1, n=20, J=40) == 6 
    assert round(effect_size(design='bcra3_2r', rho3=0.38, rho2=0.10, omega3=0.50, r21=0.37, r22=0.53, r2t3=0, g=0, n=20, J=2, K=64),3) == 0.200
    assert sample_size(design='bcra3_2r', es=0.20, rho3=0.38, rho2=0.10, omega3=0.50, r21=0.37, r22=0.53, r2t3=0, g=0, n=20, J=2) == 64
    assert round(effect_size(design='bcra4_2r', rho4=0.05, rho3=0.15, rho2=0.15, omega4=0.5, omega3=0.5, r21=0.5, r22=0.5, r2t3=0.5, r2t4=0.5, g=0, n=10, J=4, K=4, L=20), 3)==0.146
    assert sample_size(design='bcra4_2r', es=0.2, rho4=0.05, rho3=0.15, rho2=0.15, omega4=0.5, omega3=0.5, r21=0.5, r22=0.5, r2t3=0.5, r2t4=0, g=1, n=10, J=4, K=10) == 10
    assert round(effect_size(design='bcra4_3f', rho3=0.15, rho2=0.15, r21=0.5, r22=0.5, r23=0.5, g=2, n=10, J=4, K=4, L=15),3) == 0.240
    assert sample_size(design='bcra4_3f', es=0.30, rho3=0.15, rho2=0.15, r21=0.5, r22=0.5, r23=0.5, g=1, n=10, J=4, K=4) == 10
    assert round(effect_size(design='bcra4_3r', rho4=0.05, rho3=0.15, rho2=0.15, omega4=0.5, r21=0.5, r22=0.5, r23=0.5, r2t4=0.5, g=3, n=10, J=4, K=20, L=20),3) == 0.121
    assert sample_size(design='bcra4_3r', es=0.20, rho4=0.10, rho3=0.10, rho2=0.10, omega4=0.5, r21=0.5, r22=0.5, r23=0.5, r2t4=0.5, g=3, n=10, J=4, K=10) == 13
    assert round(effect_size(design = "its_nocompare", rho2=0.03, T=5, n=75, K=10, r22=0, tf=2, g=0),2) ==0.37
    assert round(effect_size(design = "its_wcompare", rho2=0.03, T=5, n=75, K=10, r22=0, tf=2, g=0, q=2),2) ==0.45
    assert sample_size(design = "its_wcompare", es=0.4, rho2=0.03, T=5, n=75, r22=0, tf=2, g=0, q=2, two_tailed=False) ==10




pypowerup/core.py
from scipy import stats
import numpy as np
from innerscope import scoped_function
import sys

class ConvergenceError(Exception):
    pass

def multiplier(alpha, v, power, two_tailed):
    """
    calculate multiplier
    v: degree of freedom
    """
    if two_tailed:
        alpha *= 0.5
    t1 = stats.t.ppf(1-alpha, v)    
    t2 = np.abs(stats.t.ppf(1-power, v))
    if power >= 0.5:
        m = t1 + t2
    else:
        m = t1 - t2
    return m

design_degree_freedom = {
    "ira": lambda: n - g - 2,
    "bira2_1c": lambda: J * (n - 1) - g - 1,
    "bira2_1f": lambda: J * (n - 2) - g,
    "rd2_1f": lambda: J * (n - 2) - g,
    "bira2_1r": lambda: J - g - 1,
    "rd2_1r": lambda: J - g - 1, 
    "bira3_1r": lambda: K - g - 1,
    "bira4_1r": lambda: L - g - 1,
    "cra2_2r": lambda: J - g - 2,
    "cra3_3r": lambda: K - g - 2,
    "cra4_4r": lambda: L - g - 2,
    "bcra3_2f": lambda: K * (J - 2) - g,
    "bcra3_2r": lambda: K - g - 1,
    "bcra4_2r": lambda: L - g - 1,
    "bcra4_3f": lambda: L * (K - 2) - g,
    "bcra4_3r": lambda: L - g - 1,
    "rdc_2r": lambda: J - g - 2,
    "rdc_3r": lambda: K - g - 2,
    "rd3_2f": lambda: K * (J - 1) - g,
    "rd3_2r": lambda: K - g - 1,
    "its_nocompare": lambda: K * T - g - 1,
    "its_wcompare": lambda: K * T - g - 1,

    }

def degree_freedom_equation(design):
    d = sys._getframe().f_back.f_locals
    calc = design_degree_freedom[design]
    func = scoped_function(calc, d)
    return func().return_value

design_sse = {
    "ira": lambda: np.sqrt((1-r21)/(p*(1-p)*n)),
    "bira2_1c": lambda: np.sqrt((1-r21)/(p*(1-p)*J*n)),
    "bira2_1f": lambda: np.sqrt((1-r21)/(p*(1-p)*J*n)),
    "rd2_1f": lambda: np.sqrt((1-r21)/(p*(1-p)*J*n))*np.sqrt(design_effect),
    "bira2_1r": lambda: np.sqrt(rho2*omega2*(1-r2t2)/J + (1-rho2)*(1-r21)/(p*(1-p)*J*n)),
    "rd2_1r": lambda: np.sqrt(rho2*omega2*(1-r2t2)/J + (1-rho2)*(1-r21)*design_effect/(p*(1-p)*J*n)), 
    "bira3_1r": lambda: np.sqrt(
                rho3*omega3*(1-r2t3)/K + 
                rho2*omega2*(1-r2t2)/(J*K) + 
                (1-rho2-rho3)*(1-r21)/(p*(1-p)*J*K*n)),
    "bira4_1r": lambda: np.sqrt(
                rho4*omega4*(1-r2t4)/L + 
                rho3*omega3*(1-r2t3)/(K*L) + 
                rho2*omega2*(1-r2t2)/(J*K*L) + 
                (1-rho2-rho3-rho4)*(1-r21)/(p*(1-p)*J*K*L*n)),
    "cra2_2r": lambda: np.sqrt(rho2*(1-r22)/(p*(1-p)*J)+(1-rho2)*(1-r21)/(p*(1-p)*J*n)),
    "cra3_3r": lambda: np.sqrt(rho3*(1-r23)/(p*(1-p)*K)+rho2*(1-r22)/(p*(1-p)*J*K)+(1-rho3-rho2)*(1-r21)/(p*(1-p)*J*K*n)), 
    "cra4_4r": lambda: np.sqrt(rho4*(1-r24)/(p*(1-p)*L)+
                               rho3*(1-r23)/(p*(1-p)*K*L)+
                               rho2*(1-r22)/(p*(1-p)*J*K*L)+
                               (1-rho4-rho3-rho2)*(1-r21)/(p*(1-p)*J*K*L*n)),
    "bcra3_2f": lambda: np.sqrt(rho2*(1-r22)/(p*(1-p)*J*K)+(1-rho2)*(1-r21)/(p*(1-p)*J*K*n)),
    "bcra3_2r": lambda: np.sqrt(rho3*omega3*(1-r2t3)/K+rho2*(1-r22)/(p*(1-p)*J*K)+(1-rho3-rho2)*(1-r21)/(p*(1-p)*J*K*n)),
    "bcra4_2r": lambda: np.sqrt(rho4*omega4*(1-r2t4)/L +
                rho3*omega3*(1-r2t3)/(K*L) +
                rho2*(1-r22)/(p*(1-p)*J*K*L) +
                (1-rho4-rho3-rho2)*(1-r21)/(p*(1-p)*J*K*L*n)),
    "bcra4_3f": lambda: np.sqrt(rho3*(1-r23)/(p*(1-p)*K*L)+
                                rho2*(1-r22)/(p*(1-p)*J*K*L)+(1-rho3-rho2)*(1-r21)/(p*(1-p)*J*K*L*n)),
    "bcra4_3r": lambda: np.sqrt(rho4*omega4*(1-r2t4)/L+rho3*(1-r23)/(p*(1-p)*K*L)+
                                rho2*(1-r22)/(p*(1-p)*J*K*L)+(1-rho4-rho3-rho2)*(1-r21)/(p*(1-p)*J*K*L*n)),
    "rdc_2r": lambda: np.sqrt(design_effect)*np.sqrt(rho2*(1-r22)/(p*(1-p)*J)+
                                                     (1-rho2)*(1-r21)/(p*(1-p)*J*n)),     
    "rdc_3r": lambda: np.sqrt(design_effect)*np.sqrt(rho3*(1-r23)/(p*(1-p)*K)+
                                                     rho2*(1-r22)/(p*(1-p)*J*K)+
                                                     (1-rho2-rho3)*(1-r21)/(p*(1-p)*J*K*n)), 
    "rd3_2f": lambda: np.sqrt(design_effect)*np.sqrt(rho2*(1-r22)/(p*(1-p)*J*K)+
                                                     (1-rho2)*(1-r21)/(p*(1-p)*J*K*n)), 
    "rd3_2r": lambda: np.sqrt(rho3*omega3*(1-r2t3)/K+ design_effect*rho2*(1-r22)/(p*(1-p)*K*J) + design_effect*(1-rho2-rho3)*(1-r21)/(p*(1-p)*J*K*n)),
    "its_nocompare": lambda: (1/np.sqrt(K))*np.sqrt(1/n+rho2*(1-r22)/(1-rho2))*np.sqrt(1+1/T+(tf+(T+1)/2)**2/(T*(T+1)*(2*T+1)/6-(T+1)**2*T/4)),
    "its_wcompare": lambda: np.sqrt(1+1/q)*(1/np.sqrt(K))*np.sqrt(1/n+rho2*(1-r22)/(1-rho2))*np.sqrt(1+1/T+(tf+(T+1)/2)**2/(T*(T+1)*(2*T+1)/6-(T+1)**2*T/4))
  

    }

def sse_equation(design):
    d = sys._getframe().f_back.f_locals
    calc = design_sse[design]
    func = scoped_function(calc, d)
    return func().return_value


design_sample_size = {
    "ira": lambda: (m/es)**2 * ((1-r21)/(p*(1-p))),
    "bira2_1c": lambda: (m/es)**2 * ((1-r21)/(p*(1-p)*n)),
    "bira2_1f": lambda: (m/es)**2 * ((1-r21)/(p*(1-p)*n)),
    "rd2_1f": lambda: (m/es)**2 * ((1-r21)/(p*(1-p)*n))*design_effect,
    "bira2_1r": lambda: (m/es)**2 * (
                rho2*omega2*(1-r2t2) + 
                (1-rho2)*(1-r21)/(p*(1-p)*n)),
    "rd2_1r": lambda: (m/es)**2 * (
                rho2*omega2*(1-r2t2) + 
                (1-rho2)*(1-r21)*design_effect/(p*(1-p)*n)), 
    "bira3_1r": lambda: (m/es)**2 * (
                rho3*omega3*(1-r2t3) + 
                rho2*omega2*(1-r2t2)/J + 
                (1-rho2-rho3)*(1-r21)/(p*(1-p)*J*n)),
    "bira4_1r": lambda: ((m/es)**2 * (
                rho4*omega4*(1-r2t4) +
                rho3*omega3*(1-r2t3)/K +
                rho2*omega2*(1-r2t2)/(K*J) + 
                (1-rho4-rho3-rho2)*(1-r21)/(p*(1-p)*K*J*n))),
    "cra2_2r": lambda: (m/es)**2*(rho2*(1-r22)/(p*(1-p)) + (1-rho2)*(1-r21)/(p*(1-p)*n)),
    "cra3_3r": lambda: (m/es)**2*(rho3*(1-r23)/(p*(1-p))+rho2*(1-r22)/(p*(1-p)*J)+(1-rho3-rho2)*(1-r21)/(p*(1-p)*J*n)), 
    "cra4_4r": lambda: (m/es)**2*(rho4*(1-r24)/(p*(1-p))+
                               rho3*(1-r23)/(p*(1-p)*K)+
                               rho2*(1-r22)/(p*(1-p)*J*K)+
                               (1-rho4-rho3-rho2)*(1-r21)/(p*(1-p)*J*K*n)),
    "bcra3_2f": lambda: (m/es)**2*(rho2*(1-r22)/(p*(1-p)*J)+(1-rho2)*(1-r21)/(p*(1-p)*J*n)),
    "bcra3_2r": lambda: (m/es)**2 * (rho3*omega3*(1-r2t3) +
                       rho2*(1-r22)/(p*(1-p)*J) +
                       (1-rho3-rho2)*(1-r21)/(p*(1-p)*J*n)),
    "bcra4_2r": lambda: (m/es)**2*(rho4*omega4*(1-r2t4)+rho3*omega3*(1-r2t3)/K+
                                   rho2*(1-r22)/(p*(1-p)*K*J)+(1-rho4-rho3-rho2)*(1-r21)/(p*(1-p)*J*K*n)),
    "bcra4_3f": lambda: (m/es)**2 * (rho3*(1-r23)/(p*(1-p)*K)+
                                     rho2*(1-r22)/(p*(1-p)*K*J)+(1-rho3-rho2)*(1-r21)/(p*(1-p)*J*K*n)),
    "bcra4_3r": lambda: (m/es)**2 * (rho4*omega4*(1-r2t4)+rho3*(1-r23)/(p*(1-p)*K)+rho2*(1-r22)/(p*(1-p)*K*J)+
                                     (1-rho4-rho3-rho2)*(1-r21)/(p*(1-p)*K*J*n)),   
    "rdc_2r": lambda: (m/es)**2*design_effect*(rho2*(1-r22)/(p*(1-p))+(1-rho2)*(1-r21)/(p*(1-p)*n)),
    "rdc_3r": lambda: (m/es)**2*design_effect*(rho3*(1-r23)/(p*(1-p))+
                                                     rho2*(1-r22)/(p*(1-p)*J)+
                                                     (1-rho2-rho3)*(1-r21)/(p*(1-p)*J*n)),
    "rd3_2f": lambda: (m/es)**2*(design_effect)*(rho2*(1-r22)/(p*(1-p)*J)+(1-rho2)*(1-r21)/(p*(1-p)*J*n)), 
    "rd3_2r": lambda: (m/es)**2*(rho3*omega3*(1-r2t3)+ design_effect*rho2*(1-r22)/(p*(1-p)*J) + design_effect*(1-rho2-rho3)*(1-r21)/(p*(1-p)*J*n)),
    "its_nocompare": lambda: (m/es)**2*(1/n+rho2*(1-r22)/(1-rho2))*(1+1/T+(tf+(T+1)/2)**2/(T*(T+1)*(2*T+1)/6-(T+1)**2*T/4)),
    "its_wcompare": lambda: (m/es)**2*(1+1/q)*(1/n+rho2*(1-r22)/(1-rho2))*(1+1/T+(tf+(T+1)/2)**2/(T*(T+1)*(2*T+1)/6-(T+1)**2*T/4))


                                }

def sample_size_equation(design):
    d = sys._getframe().f_back.f_locals
    calc = design_sample_size[design]
    func = scoped_function(calc, d)
    return func().return_value

def effect_size(design, n,  power=0.8, alpha=0.05, two_tailed=True, p=0.5, g=0, r21=0, 
         J=1, rho2=0.1, omega2 = 0.1, r2t2=0, 
         K=10, rho3=0.1, omega3 = 0.1, r2t3=0, 
         L=10, rho4=0.1, omega4 = 0.1, r2t4=0,                
         design_effect=0, r22=0, r23=0, r24=0,
         T=0, tf=0, q=0
        ):    
    #get equation for degree of freedom
    v = degree_freedom_equation(design)
    
    #get equation for sse
    sse = sse_equation(design)
    
    m = multiplier(alpha, v, power, two_tailed)
    mdes = m * sse 
    return mdes
        
def power(design, n,  es=0.25, alpha=0.05, two_tailed=True, p=.50, g=0, r21=0,
          J=1, rho2=0.1, omega2 = 0.1, r2t2=0,
          K=10, rho3=0.1, omega3 = 0.1, r2t3=0,
          L=10, rho4=0.1, omega4 = 0.1, r2t4=0,
          design_effect=0, r22=0, r23=0, r24=0,
          T=0, tf=0, q=0
         ):
    """
    calculate power 
    """
    #get equation for degree of freedom
    v = degree_freedom_equation(design)
    
    #get equation for sse
    sse = sse_equation(design)

    lambda_ = es/sse
    if two_tailed == True:
        power = (1 - stats.t.cdf(stats.t.ppf(1-alpha/2, v), v, lambda_) + 
                 stats.t.cdf(-stats.t.ppf(1-alpha/2, v), v, lambda_))
    else:
        power = 1-stats.t.cdf(stats.t.ppf(1-alpha, v), v, lambda_)
    return power

def sample_size(design, es=0.2, power=0.8, alpha=0.05, two_tailed=True, n=30, tol=.10, p=.50, g=0, r21=0, 
         J=10, rho2=0, omega2 = 0, r2t2=0, 
         K=5, rho3=0, omega3 = 0, r2t3=0, 
         L=10, rho4=0.1, omega4 = 0.1, r2t4=0,
         design_effect=0, r22=0, r23=0, r24=0, 
         T=0, tf=0, q=0,
         max_iter=500):
    """
    calculate minimum required sample sizes 
    """  
    for i in range(max_iter):
        v = degree_freedom_equation(design)
        sse = sse_equation(design)
        m = multiplier(alpha, v, power, two_tailed)   

        if design in {'bira2_1r','bira2_1c','bira2_1f','rd2_1f','rd2_1r','rdc_2r','cra2_2r'}:            
            J_new = sample_size_equation(design)
            if abs(J_new-J) < tol:
                return round((J+J_new)/2) 
            J = (J+J_new)/2            
        elif design in {'bira3_1r', 'rdc_3r', 'rd3_2f', 'rd3_2r',
                        'cra3_3r','bcra3_2f','bcra3_2r', 
                        'its_nocompare','its_wcompare'}:
            K_new = sample_size_equation(design)
            if abs(K_new-K) < tol:
                return round((K+K_new)/2) 
            K = (K+K_new)/2    

        elif design in {'bira4_1r','cra4_4r','bcra4_2r','bcra4_3f','bcra4_3r'}:
            L_new = sample_size_equation(design)
            if abs(L_new-L) < tol:
                return round((L+L_new)/2) 
            L = (L+L_new)/2  
        elif design in {'ira'}:
            n_new = sample_size_equation(design)
            # when the difference between the calculated sample size and “guestimate” is within ± 0.1 (tol).
            if abs(n_new-n) < tol:
                return round((n+n_new)/2) 
            # The “guestimate” is replaced with the average of the original “guestimate” and the calculated sample size
            n = (n+n_new)/2
    raise ConvergenceError("When calculating sample size, not converge at max iteration. You may increase max_iter.")
